### PR TITLE
Used BatchSpansProcessor instead of SimpleSpansProcessor

### LINKED ISSUE #368: Don't use synchronous span processor
Switch from `SimpleSpanProcessor` to `BatchSpansProcessor` or `DisruptorAsyncSpanProcessor`.

See https://github.com/open-telemetry/opentelemetry-auto-instr-java/blob/v0.2.2/agent-tooling/src/main/java/io/opentelemetry/auto/tooling/TracerInstaller.java#L72

### PR DESCRIPTION
**Pull request by RashmiRam**
BatchSpansProcessor won't get blocked by the exporter.
Fixes #368

@trask This is the quick fix. Are we working on introducing the disruptor in the future? If that's the case, Can this just be a stop gap for now?

### CODE REVIEW

**Comment by RashmiRam** [comment]
I signed it

**Comment by johnbley** [comment]
Hi Rashmi - 
Thanks for the patch!  It looks like this is causing the smoke-tests to fail (you can click through the "details" to see the build logs .  You can reproduce this locally by running (e.g.) `./gradlew :smoke-tests:springboot:test`.  In `SpringBootSmokeTest.groovy` we use a logging exporter and a simple log reader to check that the spans get logged.  Perhaps there's a clean way to explicitly `forceFlush` the batch or to special-case the logging processor to not be batched?

**Comment by RashmiRam** [comment]
I was trying to run failure test alone as entire instrumentation test takes hours to run.  I'll check that. Thanks for the help @johnbley

**Comment by RashmiRam** [comment]
I fixed the springboot smoke tests. Now, I see random failures in instrumentation tests(cassandra, akka-hhtp) which are passing in my local. Any help would be much appreciated.

**Comment by iNikem** [comment]
Yeah, our tests are quite flaky right now. We are working on it :) @trask can you please rerun the pipeline?

**Comment by iNikem** [comment]
@RashmiRam I see that `check` job has failed in CI with code formatting violation. Can you please run `./gradlew googleJavaFormat` as suggested by `CONTRIBUTING.md`?

**Comment by RashmiRam** [comment]
I did @iNikem and ran ./gradlew auto-tooling:verifyGoogleJavaFormat to find any pending violations. It is passing. I am not sure why the tests in CI are failing

**Comment by iNikem** [comment]
I have checked out your repo and tried to run `./gradlew auto-tooling:verifyGoogleJavaFormat`:
```
The following files are not formatted properly:

/Users/nikem/Documents/workspace/mtp/opentelemetry-auto-instr-java/agent-tooling/src/main/java/io/opentelemetry/auto/tooling/TracerInstaller.java
```
Did you commit your changes?

**Comment by RashmiRam** [comment]
Yes. I did This is the commit. https://github.com/open-telemetry/opentelemetry-auto-instr-java/pull/393/commits/c4383bab4ad139efe67f20b153078e9c95e1d2b6

**Comment by johnbley** [comment]
> Yes. I did This is the commit. [c4383ba](https://github.com/open-telemetry/opentelemetry-auto-instr-java/commit/c4383bab4ad139efe67f20b153078e9c95e1d2b6)

I've seen googleJavaFormat's cache get confused at times; try adding some whitespace, saving, and then re-running it.

**Comment by RashmiRam** [comment]
That helped @johnbley `check` job  is passing now in CircleCI. Thanks a ton.

File: smoke-tests/src/main/groovy/io/opentelemetry/smoketest/AbstractSmokeTest.groovy [hunk 1]
```diff
@@ -70,6 +70,11 @@ abstract class AbstractSmokeTest extends Specification {
     processBuilder.environment().put("JAVA_HOME", System.getProperty("java.home"))
     processBuilder.environment().put("DD_API_KEY", API_KEY)
 
+//    // Setting configuration variables of batch span processor through env vars
+//    // This config is to immediately flush a batch of 1 span with delay of 10ms
+    processBuilder.environment().put("OTEL_BSP_MAX_EXPORT_BATCH", "1")
```
Comment (johnbley) [inline smoke-tests/src/main/groovy/io/opentelemetry/smoketest/AbstractSmokeTest.groovy:75]: I like this solution a lot.  Much cleaner than anything I had suggested.

**Comment by trask** [comment]
Hi @RashmiRam, thanks so much for the PR!

It looks like our `PlaySmokeTest` smoke test is not resilient to the extra delay created by using the `BatchSpansProcessor`:

* https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-auto-instr-java/990/workflows/923aa899-852f-4974-a524-1c410e63a2fd/jobs/8224/tests
* https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-auto-instr-java/990/workflows/923aa899-852f-4974-a524-1c410e63a2fd/jobs/8226/tests

Can you try bumping the timeout at [PlaySmokeTest.groovy#L48](https://github.com/open-telemetry/opentelemetry-auto-instr-java/blob/master/smoke-tests/play/src/test/groovy/io/opentelemetry/smoketest/PlaySmokeTest.groovy#L48) and [SpringBootSmokeTest.groovy#L42](https://github.com/open-telemetry/opentelemetry-auto-instr-java/blob/master/smoke-tests/springboot/src/test/groovy/io/opentelemetry/smoketest/SpringBootSmokeTest.groovy#L42) from 10 seconds to 15 seconds?

**Comment by RashmiRam** [comment]
@trask I have updated the timeout in both play & springbok smoke tests to 20s as 15s fails at times even in my local.

File: smoke-tests/play/src/test/groovy/io/opentelemetry/smoketest/PlaySmokeTest.groovy [hunk 2]
```diff
@@ -45,7 +45,7 @@ class PlaySmokeTest extends AbstractServerSmokeTest {
     setup:
     def spanCounter = new SpanCounter(logfile, [
       (HTTP_REQUEST_SPAN): 2,
-    ], 10000)
+    ], 20000)
```
Comment (trask) [inline smoke-tests/play/src/test/groovy/io/opentelemetry/smoketest/PlaySmokeTest.groovy:48]: still getting failures on CI, hoping we just need to bump this more
```suggestion
    ], 30000)
```

File: smoke-tests/springboot/src/test/groovy/io/opentelemetry/smoketest/SpringBootSmokeTest.groovy [hunk 3]
```diff
@@ -39,7 +39,7 @@ class SpringBootSmokeTest extends AbstractServerSmokeTest {
     def spanCounter = new SpanCounter(logfile, [
       (HANDLER_SPAN): 1,
       (SERVLET_SPAN): 1,
-    ], 10000)
+    ], 20000)
```
Comment (trask) [inline smoke-tests/springboot/src/test/groovy/io/opentelemetry/smoketest/SpringBootSmokeTest.groovy:42]: still getting failures on CI, hoping we just need to bump this more
```suggestion
    ], 30000)
```

**Comment by trask** [comment]
Hey @RashmiRam, sorry we haven't merged this yet. The play smoke test is still failing in CI even after I bumped the timeout to 30 seconds. I'm looking into it now.

**Comment by trask** [comment]
The problem is that our smoke tests use the logging exporter and parse the log file which is not reliable under concurrent logging, which can happen now with the background BatchSpanProcessor thread.

Here's an example of the problem, where LOGGED_SPAN didn't end up on it's own line:

`[opentelemetry.auto.trace 2020-05-14 22:26:38:428 +0000] [application-akka.actor.default-dispatcher-4] DEBUG io.opentelemetry.auto.bootstrap.instrumentation.java.concurrent.State - Failed to set parent span because another parent span is already set io.opentelemetry.auto.bootstrap.instrumentation.java.concurrent.State@448e2484: new: io.opentelemetry.sdk.trace.RecordEventsReadableSpan@3dc393cf, old: io.opentelemetry.sdk.trace.RecordEventsReadableSpan@3dc393cfLOGGED_SPAN GET /welcome 8c4cd5a61ad2794e`

The real solution here is to improve our test harness, I made a note here: https://github.com/open-telemetry/opentelemetry-auto-instr-java/issues/298#issuecomment-629692248

For now though, the debug logging `Failed to set parent span` that is causing the interference is unnecessarily verbose, so I fixed it. I had fixed it in #318, but I had reverted my change in #362 because of https://github.com/DataDog/dd-trace-java/pull/1403#pullrequestreview-401320709. I think re-applying #318 is good anyways, and it resolves this issue for now, until we improve the test harness.

**Comment by RashmiRam** [comment]
Thank you so much for the detailed explanation on the test failures. @trask and Thanks for the merge too. It would be great if you let me know about the release date for this too.

**Comment by trask** [comment]
Releasing `v0.3.0` tomorrow (Mon, May 18) 👍

**Pull request merged**